contracts/bid/src/tree.rs

```diff
@@ -0,0 +1,248 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use super::leaf::BidLeaf;
+use crate::contract_constants::BID_TREE_DEPTH;
+use canonical::{Canon, Store};
+use canonical_derive::Canon;
+use microkelvin::{BranchMut, Nth};
+use nstack::NStack;
+use poseidon252::tree::{
+    PoseidonBranch, PoseidonMaxAnnotation, PoseidonTree, PoseidonTreeIterator,
+};
+
+/// Append-only tree structure which wraps over `PoseidonTree<BidLeaf>`
+/// responsible of the storage and the branch generation capabilities
+/// of the Bid contract.
+#[derive(Default, Debug, Clone, Canon)]
+pub struct BidTree<S>
+where
+    S: Store,
+{
+    tree: PoseidonTree<BidLeaf, PoseidonMaxAnnotation, S, BID_TREE_DEPTH>,
+}
+
+impl<S> BidTree<S>
+where
+    S: Store,
+{
+    /// Constructor for the Tree structure.
+    pub fn new() -> Self {
+        Self {
+            tree: PoseidonTree::new(),
+        }
+    }
+
+    /// Reference to the internal poseidon tree
+    ///
+    /// We don't have a mutable reference available because all its mutation
+    /// should be protected by encapsulation
+    fn inner_mut(
+        &mut self,
+    ) -> &mut PoseidonTree<BidLeaf, PoseidonMaxAnnotation, S, BID_TREE_DEPTH>
+    {
+        &mut self.tree
+    }
+
+    /// Get a bid from a provided index
+    pub fn get(&self, idx: u64) -> Option<BidLeaf> {
+        self.tree
+            .get(idx as usize)
+            .expect("Couldn't traverse the tree.")
+    }
+
+    /// Append a bid to the tree and return its index
+    ///
+    /// The index will be the last available position
+    pub fn push(&mut self, bid: BidLeaf) -> usize {
+        self.tree.push(bid).unwrap()
+    }
+
+    /// Returns a mutable refecence to the internal `Bid` that corresponds
+    /// to the given `idx`.
+    pub fn get_mut(
+        &mut self,
+        idx: u64,
+    ) -> Option<
+        BranchMut<
+            '_,
+            NStack<BidLeaf, PoseidonMaxAnnotation, S>,
+            S,
+            BID_TREE_DEPTH,
+        >,
+    > {
+        self.inner_mut()
+            .as_mut()
+            .nth_mut::<BID_TREE_DEPTH>(idx)
+            .expect("Couldn't traverse the tree")
+    }
+
+    /// Returns a poseidon branch pointing at the specific index
+    pub fn poseidon_branch(
+        &self,
+        idx: usize,
+    ) -> Option<PoseidonBranch<BID_TREE_DEPTH>> {
+        self.tree.branch(idx).unwrap()
+    }
+
+    /// Returns an iterator over the leaves of the Tree starting from a certain
+    /// block-height.
+    pub fn iter_block_height(
+        &self,
+        block_height: u64,
+    ) -> PoseidonTreeIterator<
+        BidLeaf,
+        PoseidonMaxAnnotation,
+        S,
+        u64,
+        BID_TREE_DEPTH,
+    > {
+        self.tree.iter_walk(block_height).unwrap()
+    }
+}
+
+// TODO: Fix block-height annotated search on the tree.
```

Comment:
Can we remove the commented tests directly? I assume it won't work when we update the rest of the code anyway so I would prefer removing a commented test that is outdated. Unless there is a valid reason to keep around.

Reply:
The thing is that we should not forget about implementing this test/functionallity. Will remove the test keeping the function and add an issue for it.